### PR TITLE
Maintanence

### LINKED ISSUE #4: Improve tests
- Refactor tests to remove shared state - Currently the `cachier.spec.js` tests rely on shared state between specs (`it()` statements). This is confusing, and makes the tests brittle.
- Remove tests without asserts - Some tests only exist to manipulate the shared state so that the next test can do asserts.

### LINKED ISSUE #5: Fix missing jshint config
JSHint fails if linting `cachier.spec.js` because of missing environment config in `.jshintrc`

### LINKED ISSUE #6: Add Wallaby config
Add config to run tests in Wallaby (https://wallabyjs.com/)

### LINKED ISSUE #7: Remove requirement to install grunt-cli globally to run Grunt tasks
- Add npm scripts to run Grunt tasks

### LINKED ISSUE #8: Add CircleCI build setup
- Add CircleCI build setup to run tests

### PR DESCRIPTION
**Pull request by erlendr**
This fixes #4, fixes #5, fixes #6, fixes #7, fixes #8

- Added `package-lock.json`
- Added `grunt-cli@1.2.0` to dev deps do allow running Grunt from npm scripts
- Added Wallaby config
- Refactored spec file to reset state between tests, merged tests with no asserts
- Added CircleCI config
- Added CircleCI badge to `README.md`
- Upgraded dev deps to latest, added npm scripts for various tasks
- Made Grunt use PhantomJS for Karma, fixed config bug due to Grunt upgrade
- Cleaned up karma-config, use PhantomJS as default browser
- Made JSHint ignore third party scripts
- Temporarily disabled JSHint latedef rule for spesific line in `cachier.js`. Hides scoping/hoisting issue, should be fixed
- Fixed minor linting errors

### CODE REVIEW

File: lib/cachier.js [hunk 1]
```diff
@@ -458,7 +458,9 @@ Function.prototype.bind = Function.prototype.bind || function (thisp) {
         return retVal;
     };
 
+    /*jshint -W003 */
```
Comment (cwallervand) [inline lib/cachier.js]: What are these comments ?

File: package.json [hunk 2]
```diff
@@ -22,27 +22,34 @@
     "node": ">= 0.10.0"
   },
   "scripts": {
```
Comment (cwallervand) [inline package.json]: Maybe we should add a `prepack` script that builds the application?

File: package.json [hunk 3]
```diff
@@ -22,27 +22,34 @@
...
+    "karma": "^1.7.1",
+    "karma-chrome-launcher": "^2.2.0",
+    "karma-jasmine": "^1.1.0",
+    "karma-phantomjs-launcher": "^1.0.4",
+    "karma-requirejs": "^1.1.0",
     "karma-sinon": "^1.0.4",
     "local-xmlhttprequest": "~1.5.0",
     "requirejs": "^2.1.17",
-    "sinon": "^1.14.1"
+    "sinon": "^4.0.1"
```
Comment (cwallervand) [inline package.json]: Have all the proper migration steps been done ?

File: spec/cachier.spec.js [hunk 4]
```diff
@@ -1,77 +1,114 @@
...
-            preload.init()
+            preload.init();
         }).toThrowError();
     });
 });
 
 describe("preloader manifest with one file", function() {
     var server;
-    var firstRun = true;
     var json = [{"url": "/file1.html", "hash": "bd01856bfd2065d0d1ee20c03bd3a9af"}]; //, {"url": "/file2.html", "hash": "273604bfeef7126abe1f9bff1e45126c"}, {"url": "/file3.html", "hash": "113f6696d140c167070bcc5e24791f35" }, {"url": "/file4.html", "hash": "c474a02b4b89880c88f06bb48c362cda"}];
```
Comment (cwallervand) [inline spec/cachier.spec.js]: Some commented code here

File: spec/cachier.spec.js [hunk 5]
```diff
@@ -1,77 +1,114 @@
...
-        if(firstRun) {
-            preloader.clear();
-            firstRun = false;
-        }
-
         server = sinon.fakeServer.create();
     });
 
     afterEach(function() {
+        localStorage.clear();
```
Comment (cwallervand) [inline spec/cachier.spec.js]: Do we need to clear local storage both before and after each test ? Is it maybe best to clear it after each test ?

File: spec/cachier.spec.js [hunk 6]
```diff
@@ -1,77 +1,114 @@
...
+          var fileInLocalStorageAfterSecondRun = localStorage["_test_/file1.html_bd01856bfd2065d0d1ee20c03bd3a9af"];
+          expect(fileInLocalStorageAfterSecondRun).toBe('"file1.txt"');
     });
 
-    it("should be replaced by a file with a different hash", function() {
+    it("when checkManifest called twice with different file hashes, file should be replaced", function() {
         server.respondWith('GET', "/file1.html", [ 200, {'Content-type': 'application/json'}, 'file1.txt' ]);
-        preloader.checkManifest(replaceJson);
+        preloader.checkManifest(json);
         server.respond();
```
Comment (cwallervand) [inline spec/cachier.spec.js]: Where is the `expect`statement ?

Comment (cwallervand) [reply spec/cachier.spec.js]: Aaaah..I see now that the two tests have been merged

File: lib/cachier.js (hunk 1 above)
Comment (erlendr) [reply lib/cachier.js]: See https://github.com/SparebankenVest/cachierJS/issues/10

File: package.json (hunk 3 above)
Comment (erlendr) [reply package.json]: The only Sinon functionality used is `sinon.fakeServer`. According to the migration guides covering 1.0 -> 4.0 (http://sinonjs.org/guides/), no breaking changes have been made to `fakeServer`.

File: spec/cachier.spec.js (hunk 5 above)
Comment (erlendr) [reply spec/cachier.spec.js]: Fixed

File: package.json (hunk 2 above)
Comment (erlendr) [reply package.json]: We could, but as of now the code is not available on npm 📦 :P

Comment (cwallervand) [reply package.json]: Ah :P My bad

File: spec/cachier.spec.js (hunk 4 above)
Comment (erlendr) [reply spec/cachier.spec.js]: Removed

**Comment by erlendr** [comment]
CircleCI did as well :)

**Pull request merged**